rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -255,6 +283,13 @@ void TimeSource::create_clock_sub()
 void TimeSource::destroy_clock_sub()
 {
   std::lock_guard<std::mutex> guard(clock_sub_lock_);
+  if (clock_executor_thread_.joinable()) {
+    while (!is_clock_executor_cancelled_) {
+      clock_executor_.cancel();
+    }
```

Comment:
This still has the same problem if `clock_executor_.cancel()` is callbed before `clock_executor_.spin()`. after all, to support the race condition, i think https://github.com/ros2/rclcpp/issues/1454#issuecomment-799695171 is better to give the responsibility to the `Executor` for this.

Reply:
@fujitatomoya you mean it will block forever if `create_clock_sub` is never called? yeah, is `create_clock_sub` guaranteed to be called? if so, its ok? if not, but `create_clock_sub`, if its called, is guaranteed to be called in the same thread as the destructor, then this flag can be initially set to true, an set to false only when `create_clock_sub` is called.
If that doesn't work, will need more sophisticated synchronisation machinery... which is https://github.com/ros2/rclcpp/issues/1454#issuecomment-799695171, looks like that would be a nice solution

Reply:
Ah, i see what you mean, thanks for pointing that out. this should be working eventually but it might keep calling `Executor::cancel()` and it consumes CPU for nothing. as you mentioned, probably we would want more sophisticated synchronization.

Reply:
I applied the method you recommended [here](https://github.com/ros2/rclcpp/pull/1556/commits/41eba461681f9ea8a819020dd6e7721687e8401f)